### ISSUE TITLE
Error: "No EdmType found for type..."

### ISSUE THREAD

**Issue by RubixRevenge** [moozzyk/CodeFirstFunctions#23]
Not sure if this is a bug or user error, but I have been unable to find any other examples than the one shown on this GitHub site. If it's not a bug, could someone expand the existing documentation to cover a scenario like this and its resolution?

In my SQL Server database, I have a Table Valued Function that I need to query in my MVC app (.NET Framework 4.6.1).  

Using the code sample, I wrote this in my DbContext.

```
protected override void OnModelCreating(DbModelBuilder modelBuilder)
{
    modelBuilder.Conventions.Add(new FunctionsConvention<MiaFy19Context>("dbo"));
}

public ObjectResult<Mia1PartASummaryBar> GetMia1PartASummaryBars()
{
    return ((IObjectContextAdapter)this).ObjectContext.
        ExecuteFunction<Mia1PartASummaryBar>("Mia1A_SummaryBar");

}
```

However, I received this error viewing in the browser. (Line 61 was highlighted in red.)

```
Exception Details: System.InvalidOperationException: 
       No EdmType found for type 'MiaFy19.Core.Models.Mia1PartASummaryBar'.

Line 59:         {
Line 60: 
Line 61:             return ((IObjectContextAdapter)this).ObjectContext.
Line 62:                 ExecuteFunction<Mia1PartASummaryBar>("Mia1A_SummaryBar");
Line 63: 
```

In my SQL Server database, this is what the function signature looks like.

```
ALTER FUNCTION [dbo].[Mia1A_SummaryBar] ()
RETURNS TABLE
AS
RETURN (
// code...
```

Any help is appreciated. Thank you.

**Comment by moozzyk** [comment]
The error indicates that the `Mia1PartASummaryBar` is not part of your model. Can you show your model/context? 

The exception is thrown from [here](https://github.com/moozzyk/CodeFirstFunctions/blob/2e7342afcd492598f60dd7e30e1ad5c980dfdeab/CodeFirstStoreFunctions/FunctionDiscovery.cs#L222).

**Comment by RubixRevenge** [comment]
Here is my model.

```
namespace MiaFy19.Core.Models
{
    public class Mia1PartASummaryBar
    {
        public int Id { get; set; }

        [StringLength(4)]
        public string Campus { get; set; }

        [StringLength(3)]
        public string MetGoal { get; set; }
        public int Points { get; set; }
        
        public double ShrCreditRate { get; set; }

        public double NonShrCreditRate { get; set; }
    }
}
```

And here is my context.  

```
namespace MiaFy19.Persistence
{
    public class MiaFy19Context : DbContext, IMiaFy19Context
    {
        public MiaFy19Context()
            : base("name=MiaFy19Context")
        {
        }

        #region OtherDbSets
        /* .. lots of dbsets ...*/
        #endregion

        public virtual DbSet<Mia1PartASummaryBar> Mia1PartASummaryBars { get; set; }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            modelBuilder.Conventions.Add(new FunctionsConvention<MiaFy19Context>("dbo"));
        }

        [DbFunction("MiaFy19Context", "GetMia1PartASummaryBars")]
        public IQueryable<Mia1PartASummaryBar> GetMia1PartASummaryBars()
        {
            return ((IObjectContextAdapter) this).ObjectContext
                .CreateQuery<Mia1PartASummaryBar>("[Mia1A_SummaryBar]()");
        }
    }
}
```

Since I posted my question, I've added a dbset (shown below) and tried the `.CreateQuery` example. Still not working. 

This time, I get 

> 'Mia1A_SummaryBar' cannot be resolved into a valid type or function. Near escaped identifier, line 1, column 1.

```
Line 24:         public IEnumerable<Mia1PartASummaryBar> GetMia1ASummaryBar(string campus)
Line 25:         {
Line 26:             return _context.GetMia1PartASummaryBars().Where(m => m.Campus == campus).ToList();
Line 27:         }
Line 28:     }
```

I seem to be making progress in the fact I'm getting a different error now, but still no resolution.

**Comment by moozzyk** [comment]
I think the parameter to `CreateQuery` needs to be prepended with context name. If you look at README you will find that the query is created like this (notice `string.Format` and `GetType.Name().Name` used as the first parameter):

```C#
 return ((IObjectContextAdapter)this).ObjectContext
            .CreateQuery<Customer>(
                string.Format("[{0}].{1}", GetType().Name, 
                    "[CustomersByZipCode](@ZipCode)"), zipCodeParameter);
```

**Comment by RubixRevenge** [comment]
The problem is that my function has no parameters. Maybe I'm misunderstanding, but I thought that extra code was for inserting a parameter into the function call.
```
ALTER FUNCTION [dbo].[Mia1A_SummaryBar] ()
RETURNS TABLE
AS
RETURN (
```

**Comment by RubixRevenge** [comment]
So I tried this.

```
[DbFunction("MiaFy19Context", "GetMia1PartASummaryBars")]
        public IQueryable<Mia1PartASummaryBar> GetMia1PartASummaryBars()
        {
            return ((IObjectContextAdapter) this).ObjectContext
                .CreateQuery<Mia1PartASummaryBar>(string.Format("[{0}].{1}".GetType().Name, "[Mia1A_SummaryBar]()"));
        }
```

Slightly different error this time. 

```
'String' could not be resolved in the current scope or context. Make sure that 
all referenced variables are in scope, that required schemas are loaded, and that 
namespaces are referenced correctly. Near simple identifier, line 1, column 1.
Description: An unhandled exception occurred during the execution of the current 
web request. Please review the stack trace for more information about the error 
and where it originated in the code. 

Exception Details: System.Data.Entity.Core.EntitySqlException: 'String' could 
not be resolved in the current scope or context. Make sure that all referenced 
variables are in scope, that required schemas are loaded, and that namespaces are 
referenced correctly. Near simple identifier, line 1, column 1.

Source Error: 

Line 24:         public IEnumerable<Mia1PartASummaryBar> GetMia1ASummaryBar(string campus)
Line 25:         {
Line 26:             return _context.GetMia1PartASummaryBars().Where(m => m.Campus == campus).ToList();
Line 27:         }
Line 28:     }
```

**Comment by moozzyk** [comment]
I think there is a typo in your code. Inside string.Format you put
.GetType... (notice the dot) instead of, GetType... The result is that you
are passing 'String' as the name of the function instead of the expected
fully qualified name of the function. Once you change '.' to ',' string.
Format should pick the context type name as the parameter and create a
fully qualified man that you need (you could actually use string
interpolation to make it more readable).
It is OK to not have parameters. This part looks good to me.

**Comment by RubixRevenge** [comment]
> Even changing the dot to a comma, I'm still getting the error.
> MiaFy19Context.Mia1A_SummaryBar' cannot be resolved into a valid type or function. Near member access expression, line 1, column 17.

```
[DbFunction("MiaFy19Context", "GetMia1PartASummaryBars")]
public IQueryable<Mia1PartASummaryBar> GetMia1PartASummaryBars()
{
	return ((IObjectContextAdapter) this).ObjectContext
		.CreateQuery<Mia1PartASummaryBar>(string.Format("[{0}].{1}", 
		GetType().Name, "[Mia1A_SummaryBar]()"));
}
```

**Comment by RubixRevenge** [comment]
Got it worked out. :)

Although an `Id` field is meaningless on the data-side, it turns out that EF needs it, so I made the following transformations. It think it's not just one step, but the steps in combination that did the trick.

1. Added an Id column in my SQL TVF and gave it unique value.
2. In my entity class, I added the `[Key]` attribute -- just in case it helped.
3. I changed the argument of `CreateQuery()`. 

```
[DbFunction("MiaFy19Context", "Mia1A_SummaryBar")]
public virtual IQueryable<Mia1ASummaryBar> Mia1A_SummaryBar()
{
	return ((IObjectContextAdapter)this).ObjectContext
		.CreateQuery<Mia1ASummaryBar>("MiaFy19Context.[Mia1A_SummaryBar]()");
}
```

**Comment by RubixRevenge** [comment]
I'm going to mark this issue as closed as my issue is resolved. For the sake of any other first-time users, it might be helpful to add some "pitfalls, gotchas, assumptions, and resolutions" items to the documentation.

**Comment by moozzyk** [comment]
I am glad you were able to make it work. The rows returned from the TVF have to match the result type  - the `Mia1PartASummaryBar` entity has the Id property so the result returned from TVF should also have it. Store functions in EF (be it stored procs or TVFs) need to return types that EF knows about (i.e. are part of the model). This can be achieved either by creating a DbSet or just by adding the type in `OnModelCreating`. The sample from README shows how to bind a TVFs (in retrospect - the sample should use string interpolation but I think at the time I wrote the code string interpolation was not supported). The README also contains links to post I wrote: 
- https://blog.3d-logic.com/2014/08/11/the-beta-version-of-store-functions-for-entityframework-6-1-1-code-first-available/ 
- https://blog.3d-logic.com/2014/10/18/the-final-version-of-the-store-functions-for-entityframework-6-1-1-code-first-convention-released/
